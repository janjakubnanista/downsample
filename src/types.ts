/**
 * Possible types for the X coordinate (most probably time)
 */
export type X = number | Date;

/**
 * Possible types for the Value coordinate
 */
export type Value = number;

export type NormalizedDataPoint = [Value, Value];

export type TupleDataPoint = [X, Value];

export interface XYDataPoint {
  x: X;
  y: Value;
}

export type DataPoint = TupleDataPoint | XYDataPoint;

export type NumericPropertyNames<T> = {
  [K in keyof T]: T[K] extends Value ? K : never;
}[keyof T];

// TODO P extends [] does not cover all the tuple cases with heterogenous tuples
// and it would be nice to have a (generated) type for that
export type NumericPropertyAccessor<P> = P extends unknown[] ? number : NumericPropertyNames<P>;

export type PointValueExtractor<P> = (point: P) => Value;

export interface DownsamplingFunctionConfig<P> {
  x: NumericPropertyAccessor<P> | PointValueExtractor<P>;
  y: NumericPropertyAccessor<P> | PointValueExtractor<P>;
}

export interface SmoothingFunctionConfig<P> extends DownsamplingFunctionConfig<P> {
  toPoint: (x: number, y: number, index: number) => P;
}

export type DownsamplingFunction<I, A extends unknown[], O = I> = (data: I[], ...args: A) => O[];
