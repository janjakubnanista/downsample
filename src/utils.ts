import { NormalizedDataPoint } from './types';
import { isA } from 'ts-type-checked';

export function calculateTriangleArea(
  pointA: NormalizedDataPoint,
  pointB: NormalizedDataPoint,
  pointC: NormalizedDataPoint,
): number {
  return Math.abs((pointA.x - pointC.x) * (pointB.y - pointA.y) - (pointA.x - pointB.x) * (pointC.y - pointA.y)) / 2;
}

export function calculateAverageDataPoint(...points: NormalizedDataPoint[]): NormalizedDataPoint | undefined {
  const { length } = points;
  if (!length) return undefined;

  let x = 0;
  let y = 0;
  for (let i = 0; i < length; i++) {
    x += points[i].x;
    y += points[i].y;
  }

  return { x: x / length, y: y / length };
}

export function splitIntoBuckets<T>(data: T[], desiredLength: number): T[][] {
  if (data.length === 2) {
    return [[data[0]], [data[1]]];
  }

  const first: T = data[0];
  const center: T[] = data.slice(1, data.length - 1);
  const last: T = data[data.length - 1];

  // First and last bucket are formed by the first and the last data points
  // so we only have N - 2 buckets left to fill
  const bucketSize: number = center.length / (desiredLength - 2);
  const splitData: T[][] = [[first]];

  for (let i = 0; i < desiredLength - 2; i++) {
    const bucketStartIndex: number = Math.floor(i * bucketSize);
    const bucketEndIndex: number = Math.floor((i + 1) * bucketSize);
    const dataPointsInBucket: T[] = center.slice(bucketStartIndex, bucketEndIndex);

    splitData.push(dataPointsInBucket);
  }

  splitData.push([last]);

  return splitData;
}

export const calculateMean = (values: number[]): number => {
  let m = 0;
  for (let i = 0; i < values.length; i += 1) {
    m += values[i];
  }

  return m / values.length;
};

export const calculateSTD = (values: number[]): number => {
  const mean = calculateMean(values);

  let std = 0;
  for (let i = 0; i < values.length; i += 1) {
    std += (values[i] - mean) * (values[i] - mean);
  }

  return Math.sqrt(std / values.length);
};

// export const createNormalize = <P>(
//   x: NumericPropertyAccessor<P> | PointValueExtractor<P>,
//   y: NumericPropertyAccessor<P> | PointValueExtractor<P>,
// ) => {
//   const getX = getPointValueExtractor(x);
//   const getY = getPointValueExtractor(y);

//   return (data: PossibleArray<P>): NormalizedDataPoint[] => data.map((point: P, index: number) => [getX(point, index), getY(point, index)]);
// };

// export const createXYDataPoint = (time: number, value: number): XYDataPoint => ({ x: time, y: value });

// export const createLegacyDataPointConfig = (): SmoothingFunctionConfig<DataPoint> => ({
//   x: (point: DataPoint) => {
//     const t = isA<XYDataPoint>(point) ? point.x : point[0];

//     return isA<Date>(t) ? t.getTime() : t;
//   },
//   y: (point: DataPoint) => ('y' in point ? point.y : point[1]),
//   toPoint: createXYDataPoint,
// });
