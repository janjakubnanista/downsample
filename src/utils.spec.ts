import 'jest';
import { NormalizedDataPoint } from './types';
import { calculateAverageDataPoint, calculateTriangleArea, splitIntoBuckets } from './utils';

describe('utils', () => {
  describe('calculateTriangleArea', () => {
    it('should return 0 for a collapsed triangle', () => {
      const pointA: NormalizedDataPoint = [1, 2];
      const pointB: NormalizedDataPoint = [2, 2];
      const pointC: NormalizedDataPoint = [3, 2];

      expect(calculateTriangleArea(pointA, pointB, pointC)).toBe(0);
    });

    it('should return correct area for a triangle aligned with an axis', () => {
      const pointA: NormalizedDataPoint = [1, 0];
      const pointB: NormalizedDataPoint = [2, 5];
      const pointC: NormalizedDataPoint = [3, 0];

      expect(calculateTriangleArea(pointA, pointB, pointC)).toBe(5);
    });

    it('should return correct area for a triangle not aligned with an axis', () => {
      const pointA: NormalizedDataPoint = [1, 2];
      const pointB: NormalizedDataPoint = [2, 5];
      const pointC: NormalizedDataPoint = [3, 5];

      expect(calculateTriangleArea(pointA, pointB, pointC)).toBe(1.5);
    });

    it('should return correct area for a triangle covering negative quadrants', () => {
      const pointA: NormalizedDataPoint = [-1, -1];
      const pointB: NormalizedDataPoint = [0, 1];
      const pointC: NormalizedDataPoint = [1, -1];

      expect(calculateTriangleArea(pointA, pointB, pointC)).toBe(2);
    });
  });

  describe('calculateAverageDataPoint', () => {
    it('should return undefined when passed no data points', () => {
      expect(calculateAverageDataPoint()).toBe(undefined);
    });

    it('should return the same point when passed one data point', () => {
      expect(calculateAverageDataPoint([0, 1])).toEqual([0, 1]);
    });

    it('should return correct point average', () => {
      const pointA: NormalizedDataPoint = [-1, -1];
      const pointB: NormalizedDataPoint = [0, 2];
      const pointC: NormalizedDataPoint = [1, -1];

      expect(calculateAverageDataPoint(pointA, pointB, pointC)).toEqual([0, 0]);
    });
  });

  describe('spliIntoBuckets', () => {
    const data: [number, number][] = [
      [0, 2],
      [1, -1],
      [2, 2],
      [3, -1],
      [4, 2],
      [5, -1],
      [6, 2],
      [7, -1],
    ];

    it('should return two buckets with one data point each when passed two data points', () => {
      expect(
        splitIntoBuckets(
          [
            [0, 2],
            [1, -1],
          ],
          2,
        ),
      ).toEqual([[[0, 2]], [[1, -1]]]);
    });

    it('should return an array of desired length', () => {
      expect(splitIntoBuckets(data, 2)).toHaveLength(2);
      expect(splitIntoBuckets(data, 3)).toHaveLength(3);
      expect(splitIntoBuckets(data, 4)).toHaveLength(4);
      expect(splitIntoBuckets(data, 5)).toHaveLength(5);
    });

    it('should return an array with the first and the last bucket containing the first and the last data points', () => {
      expect(splitIntoBuckets(data, 3)[0]).toHaveLength(1);
      expect(splitIntoBuckets(data, 3)[0]).toEqual([[0, 2]]);
      expect(splitIntoBuckets(data, 3)[2]).toHaveLength(1);
      expect(splitIntoBuckets(data, 3)[2]).toEqual([[7, -1]]);
    });
  });
});
