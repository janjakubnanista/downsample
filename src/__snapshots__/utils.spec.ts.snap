// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`utils calculateSMA should return okay numbers 1`] = `
Array [
  1.8,
  2.2,
  3,
  3.8,
  4.2,
]
`;

exports[`utils calculateSMA should return okay numbers 2`] = `
Array [
  1,
  1,
  2,
  3,
  3,
  4,
  5,
  5,
]
`;

exports[`utils calculateSMA should return okay numbers 3`] = `
Array [
  1.5,
  2,
  2.5,
  3.5,
  4,
  4.5,
]
`;

exports[`utils calculateSMA should return okay numbers 4`] = `
Array [
  1,
  3,
  5,
  7,
  9,
]
`;

exports[`utils calculateSMA should return okay numbers 5`] = `
Array [
  1.5,
  4.5,
  7.5,
]
`;

exports[`utils calculateSMA should return okay numbers 6`] = `
Array [
  1,
  5,
  9,
]
`;

exports[`utils calculateSMA should return okay numbers 7`] = `
Array [
  1,
  3,
  5,
  7,
  9,
]
`;

exports[`utils calculateSMA should return okay numbers 8`] = `
Array [
  1.5,
  4.5,
  7.5,
]
`;

exports[`utils calculateSMA should return okay numbers 9`] = `
Array [
  2,
  6,
]
`;
